exchanges/dydx/dydx_websocket.go

```diff
@@ -0,0 +1,328 @@
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/account"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/ticker"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/trade"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const (
+
+	// channels
+
+	accountsChannel  = "v3_accounts"
+	orderbookChannel = "v3_orderbook"
+	tradesChannel    = "v3_trades"
+	marketsChannel   = "v3_markets"
+)
+
+var defaultSubscriptions = []string{
+	orderbookChannel,
+	tradesChannel,
+	marketsChannel,
+}
+
+// WsConnect connect to dydx websocket server.
+func (dy *DYDX) WsConnect() error {
+	if !dy.Websocket.IsEnabled() || !dy.IsEnabled() {
+		return errors.New(stream.WebsocketNotEnabled)
+	}
+	var dialer websocket.Dialer
+	err := dy.Websocket.Conn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	dy.Websocket.Conn.SetupPingHandler(stream.PingHandler{
+		Message:     []byte(`pong`),
+		MessageType: websocket.TextMessage,
+		Delay:       time.Second * 30,
+	})
+	if dy.Verbose {
+		log.Debugf(log.ExchangeSys, "%s Connected to Websocket.\n", dy.Name)
+	}
+	dy.Websocket.Wg.Add(1)
+	go dy.wsReadData()
+	return nil
+}
+
+func (dy *DYDX) wsReadData() {
+	defer dy.Websocket.Wg.Done()
+	for {
+		resp := dy.Websocket.Conn.ReadMessage()
+		if resp.Raw == nil {
+			return
+		}
+		err := dy.wsHandleData(resp.Raw)
+		if err != nil {
+			dy.Websocket.DataHandler <- err
+		}
+	}
+}
+
+func (dy *DYDX) wsHandleData(respRaw []byte) error {
+	var resp WsResponse
+	err := json.Unmarshal(respRaw, &resp)
+	if err != nil {
+		return err
+	}
+	switch resp.Channel {
+	case accountsChannel:
+		switch resp.Type {
+		case "subscribe":
+			var resp AccountSubscriptionResponse
+			err = json.Unmarshal(respRaw, &resp)
+			if err != nil {
+				return err
+			}
+			err = dy.processOrders(resp.Contents.Orders)
+			if err != nil {
+				return err
+			}
+			dy.Websocket.DataHandler <- resp.Transfers
+			dy.Websocket.DataHandler <- resp.FundingPayments
+		case "channel_data":
+			var resp AccountChannelData
+			err := json.Unmarshal(respRaw, &resp)
+			if err != nil {
+				return err
+			}
+			err = dy.processOrders(resp.Contents.Orders)
+			if err != nil {
+				return err
+			}
+			dy.Websocket.DataHandler <- resp.Contents.Accounts
+			dy.Websocket.DataHandler <- resp.Contents.Positions
+			dy.Websocket.DataHandler <- resp.Contents.Fills
+		}
+		return nil
+	case orderbookChannel:
+		var market MarketOrderbook
+		err := json.Unmarshal(respRaw, &market)
```

Comment:
possible shadow declaration issue, please use assignment operators across this block and below as our linters will eventually pick them up when they improve.